src/tbp/interactive/tutorials/simple_visualization/tutorial.md

```diff
@@ -0,0 +1,888 @@
+
+        self._add_kwargs = {
+            "xmin": 0,
+            "xmax": 10,
+            "value": 0,
+            "pos": [(0.1, 0.2), (0.9, 0.2)],
+            "title": "Episode",
+        }
+
+        self._locators = self.create_locators()
+
+    def create_locators(self) -> dict[str, DataLocator]:
+        locators = {}
+        locators["episode"] = DataLocator(
+            path=[DataLocatorStep.key(name="episode")],
+        )
+        return locators
+
+    def add(self, callback: Callable[[Slider2D, str], None]) -> Slider2D:
+        kwargs = deepcopy(self._add_kwargs)
+        locator = self._locators["episode"]
+        kwargs.update({"xmax": len(self.data_parser.query(locator)) - 1})
+
+        widget = self.plotter.at(0).add_slider(callback, **kwargs)
+        self.plotter.at(0).render()
+        return widget
+
+    def extract_state(self, widget: Slider2D) -> int:
+        return extract_slider_state(widget)
+
+    def set_state(self, widget: Slider2D, value: int) -> None:
+        set_slider_state(widget, value)
+
+    def state_to_messages(self, state: int) -> Iterable[TopicMessage]:
+        return [TopicMessage(name="episode_number", value=state)]
+```
+
+The most important method here is `state_to_messages()`.
+This is where widget state becomes part of the pub/sub system.
+Every time the slider changes, the Widget wrapper asks `EpisodeSliderWidgetOps` for messages representing the new state.
+In this case, the widget publishes a single message named `episode_number` whose value is the selected episode index.
+
+The `add()` method is also important.
+Instead of hard-coding the slider range, it queries the experiment log for available episodes and sets `xmax` accordingly.
+This keeps the UI consistent with the data, even if the number of episodes changes between runs.
+
+At the end of this part, the slider should be visible in renderer 0, and moving it publishes `episode_number` messages.
+See the animation below.
+
+
+![Episode interactive plot](assets/episode.gif)
+
+
+## Part 3: Add a Step Slider That Reacts to the Episode Slider
+
+In Part 2 we added the episode slider and saw how a widget can publish state as a topic message.
+In this part we add the step slider, and this is the first time we introduce subscriptions.
+The step slider depends on the currently selected episode because each episode can have a different number of valid steps.
+That means the step slider must listen for messages on the topic `episode_number` and adjust its range whenever the episode changes.
+
+We will follow the same pattern as before.
+First we create the widget in `InteractivePlot.create_widgets()`.
+Then we implement the `StepSliderWidgetOps` class.
+The new concept is `WidgetUpdater`, which lets a widget react to one or more topics and update itself in response.
+
+
+### Add the Step Slider Widget to `InteractivePlot`
+
+Start by adding the new imports for `WidgetUpdater`, `TopicSpec`, and any types needed by the step slider.
```

Comment:
"any types needed by the step slider" - I don't see any types imported, so is this comment necessary? I think if you're referring to requirements outside the tutorial, then I wouldn't mention it, since that should be fairly self-explanatory.